src/base/associative_cache.hh

```diff
@@ -0,0 +1,308 @@
+#define __BASE_ASSOCIATIVE_CACHE_HH__
+
+#include <vector>
+
+#include "base/cache_entry.hh"
+#include "base/intmath.hh"
+#include "base/named.hh"
+#include "mem/cache/replacement_policies/base.hh"
+#include "mem/cache/replacement_policies/replaceable_entry.hh"
+#include "mem/cache/tags/indexing_policies/base.hh"
+#include "types.hh"
+
+namespace gem5
+{
+
+/** Cache Library.
+ *
+ */
+template <typename Entry>
+class AssociativeCache : public Named
+{
+    static_assert(std::is_base_of_v<CacheEntry, Entry>,
+                  "Entry should be derived from CacheEntry");
+
+    typedef replacement_policy::Base BaseReplacementPolicy;
+
+  protected:
+
+    /** The number of entries in the cache. */
+    size_t numEntries;
+
+    /** Associativity of the cache. */
+    size_t associativity;
+
+    /** The replacement policy of the cache. */
+    BaseReplacementPolicy *replPolicy;
+
+    /** Indexing policy of the cache */
+    BaseIndexingPolicy *indexingPolicy;
+
+    /** The entries */
+    std::vector<Entry> entries;
+
+  private:
+
+    void initParams(size_t _num_entries, size_t _assoc)
+    {
+        fatal_if(!isPowerOf2(_num_entries), "The number of entries of an "
+                 "AssociativeCache<> must be a power of 2");
+        fatal_if(!isPowerOf2(_assoc), "The associativity of an "
+                 "AssociativeCache<> must be a power of 2");
+        for (auto entry_idx = 0; entry_idx < _num_entries; entry_idx++) {
+            Entry *entry = &entries[entry_idx];
+            indexingPolicy->setEntry(entry, entry_idx);
+            entry->replacementData = replPolicy->instantiateEntry();
+        }
+    }
+
+  public:
+
+    /**
+     * Empty constructor - need to call init() later with all args
+     */
+    AssociativeCache(const char *name) : Named(std::string(name)) {}
+
+    /**
+     * Public constructor
+     * @param name Name of the cache
+     * @param num_entries total number of entries of the container, the number
+     *   of sets can be calculated dividing this balue by the 'assoc' value
+     * @param associativity number of elements in each associative set
+     * @param entry_size Size of each entry
+     * @param repl_policy replacement policy
+     * @param indexing_policy indexing policy
+     */
+    AssociativeCache(const char *name, const size_t num_entries,
+                     const size_t associativity_,
+                     BaseReplacementPolicy *repl_policy,
+                     BaseIndexingPolicy *indexing_policy,
+                     Entry const &init_val = Entry())
+        : Named(std::string(name)), numEntries(num_entries),
+          associativity(associativity_),
+          replPolicy(repl_policy),
+          indexingPolicy(indexing_policy),
+          entries(num_entries, init_val)
+    {
+        initParams(numEntries, associativity);
+    }
+
+    ~AssociativeCache() = default;
+
+    /**
+     * Clear the entries in the cache.
+     */
+    void clear() {
```

Comment:
The following is not a strong opinion, so you can opt to ignore the suggested changes:

clear, in std, typically (always?) means removing all elements from the container - i.e., zeroing its size. Ideally we should mimic std´s nomenclature, as users are used to it, so the learning curve is smoother. I would rename it as invalidateBlocks or something instead. However, we already use ¨insert¨ in a different fashion from std, so I guess it doesn´t matter that much in the end.

In mem/cache/tags/base you will find a forEachBlk function. It applies a function to all blocks. I think it could be better to have an equivalent iterator function here that the user could apply any function to all entries, including invalidation.